src/core/qlist.h

```diff
@@ -0,0 +1,76 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+extern "C" {
+#include "redis/quicklist.h"
+}
+
+#include <functional>
+#include <optional>
+#include <string>
+
+namespace dfly {
+
+class QList {
+ public:
+  enum Where { TAIL, HEAD };
+
+  struct Entry {
+    Entry(const char* value, size_t length) : value{value}, length{length} {
+    }
+    Entry(long long longval) : value{nullptr}, longval{longval} {
+    }
+
+    const char* value;
+    union {
+      size_t length;
+      long long longval;
+    };
+  };
+
+  using IterateFunc = std::function<bool(Entry)>;
+  enum InsertOpt { BEFORE, AFTER };
+
+  QList();
+  QList(int fill, int compress);
+  ~QList();
+
+  size_t Size() const {
+    return count_;
+  }
+
+  void Push(std::string_view value, Where where);
+  void AppendListpack(unsigned char* zl);
+  void AppendPlain(unsigned char* zl);
+  void Insert(std::string_view pivot, std::string_view elem, InsertOpt opt);
+
+  size_t MallocUsed() const;
+
+  std::string Peek(Where where) const;
+  std::optional<std::string> Get(long index) const;
```

Comment:
I would prefer to see the standard approach without optional and std::advance

Reply:
I copy the interface 1:1 to how it's been designed. After refactoring we can improve.